src/Symfony/Component/Routing/Generator/UrlGenerator.php

```diff
@@ -294,6 +294,15 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
             return $a == $b ? 0 : 1;
         });
 
+        array_walk_recursive($extra, $caster = static function (&$v) use (&$caster) {
+            if ($v instanceof \stdClass) {
+                $v = (array) $v;
+                array_walk_recursive($v, $caster);
+            } elseif (\is_object($v)) {
+                $v = (string) $v;
```

Comment:
What about array-like objects here? I'd guess that Iterators/ArrayAccess implementations would want to behave similar to `(array) $v` in the above case. What should happen in the case of an Iterator or ArrayAccess (for example) where the object can be converted to a string representation? 

Would it be an idea to give users control over how the parameters should be rendered, perhaps through an interface or custom parameter object that can deal with this?

Reply:
For those, we should mimik what http_build_query does, nothing more. A test case would help know.